### PR TITLE
Set the default path_info to the root '/'

### PR DESCRIPTION
**Pull request by zhaoguixu**
Set the default path_info to the root '/' when both the path_info passed to 'bind' and 'match' are None.
#740

### CODE REVIEW

**Comment by untitaker** [comment]
@zhaoguixu Those tests fail for me. Could you take a look? Thanks!

**Comment by zhaoguixu** [comment]
This should be good. @untitaker

**Comment by untitaker** [comment]
Please run `make stylecheck` on your code, some things are not conforming to the styleguide.

Will review content later.

File: werkzeug/routing.py [hunk 1]
```diff
@@ -1498,7 +1498,7 @@ def match(self, path_info=None, method=None, return_rule=False,
         """
         self.map.update()
         if path_info is None:
-            path_info = self.path_info
+            path_info = self.path_info or u'/'
```
Comment (untitaker) [inline werkzeug/routing.py]: Perhaps this logic would be better suited for `bind`, such that `self.path_info` is not None?

Something like:

```diff
diff --git a/werkzeug/routing.py b/werkzeug/routing.py
index 82cf793..888ac9b 100644
--- a/werkzeug/routing.py
+++ b/werkzeug/routing.py
@@ -1232,6 +1232,8 @@ class Map(object):
             subdomain = self.default_subdomain
         if script_name is None:
             script_name = '/'
+        if path_info is None:
+            path_info = '/'
         server_name = _encode_idna(server_name)
         return MapAdapter(self, server_name, script_name, subdomain,
                           url_scheme, path_info, default_method, query_args)
@@ -1498,7 +1500,7 @@ class MapAdapter(object):
         """
         self.map.update()
         if path_info is None:
-            path_info = self.path_info or u'/'
+            path_info = self.path_info
         else:
             path_info = to_unicode(path_info, self.map.charset)
         if query_args is None:
```

**Comment by untitaker** [comment]
I don't really understand why you need this behavior in the first place. In which situation would you call `match` without arguments?

**Comment by zhaoguixu** [comment]
Both the default value of ```path_info``` of method ```bind``` and ```match``` is None. This give us a sense None is acceptable. However, at most one of the two can accept None.

**Comment by untitaker** [comment]
I think it was intended to just have no implicit default at all. That this is expressed with a 404 is unfortunate.

**Comment by zhaoguixu** [comment]
I am not very familiar with the whole logic. I just test the two combined in one situation and it gives me a trackback. Feel free to decide whether this fix is necessary.

**Comment by untitaker** [comment]
Delegating to @mitsuhiko, because I don't really know.

**Comment by davidism** [comment]
continued in #1316

**Pull request closed without merging**